lib/cyrusdb_twom.c

```diff
@@ -0,0 +1,2596 @@
+        = ntohll(*((uint64_t *)(base + OFFSET_NUM_COMMITS)));
+
+    header->dirty_size
+        = ntohll(*((uint64_t *)(base + OFFSET_DIRTY_SIZE)));
+
+    header->repack_size
+        = ntohll(*((uint64_t *)(base + OFFSET_REPACK_SIZE)));
+
+    header->current_size
+        = ntohll(*((uint64_t *)(base + OFFSET_CURRENT_SIZE)));
+
+    header->maxlevel
+        = ntohl(*((uint32_t *)(base + OFFSET_MAXLEVEL)));
+
+    if (db->nocsum) return 0;
+
+    uint32_t csum = ntohl(*((uint32_t *)(base + OFFSET_CSUM)));
+    if (file->csum(base, OFFSET_CSUM) != csum) {
+        xsyslog(LOG_ERR, "DBERROR: twom header checksum failure",
+                         "filename=<%s>",
+                         db->fname);
+        return CYRUSDB_IOERROR;
+    }
+
+    return 0;
+}
+
+static size_t tm_roundup(size_t offset)
+{
+    size_t page_size = 1<<14; // 16k
+    return ((offset + offset / 4) + page_size - 1) & ~(page_size - 1);
+}
+
+/* we keep open ONLY the current file 
+ * and any others which still have a reference to them */
+
+static void _remove_txn(struct txn **ptr)
+{
+    struct txn *this = *ptr;
+    struct txn *next = this->next;
+    this->file->refcount--;
+    free(this);
+    *ptr = next;
+}
+
+static void _remove_file(struct tm_file **ptr)
+{
+    struct tm_file *this = *ptr;
+    struct tm_file *next = (*ptr)->next;
+    assert(!this->refcount);
+    assert(!this->has_lock);
+    if (this->base) munmap(this->base, this->size);
+    if (this->fd != -1) close(this->fd);
+    free(this);
+    *ptr = next;
+}
+
+static void empty_db(struct dbengine *db)
+{
+    if (db->loc.file) {
+        db->loc.file->refcount--;
+        db->loc.file = NULL;
+    }
+    while (db->write_txn)
+        _remove_txn(&db->write_txn);
+    while (db->read_txn)
+        _remove_txn(&db->read_txn);
+    while (db->openfile)
+        _remove_file(&db->openfile);
+}
+
+static void tm_cleanup(struct dbengine *db)
+{
+    // close any open files with a zero refcount (no loc or txn
+    // still points there) except the first one
+    if (!db->openfile) return;
+    struct tm_file **ptr = &db->openfile->next;
+    while (*ptr) {
+        if (!(*ptr)->refcount) {
+            _remove_file(ptr);
+        }
+        else {
+            ptr = &((*ptr)->next);
+        }
+    }
+}
+
+/* end cleanup */
+
+static inline int tm_commit(struct dbengine *db, size_t len)
+{
+    assert(db->openfile);
+    if (!db->openfile->dirty) return 0;
+    assert(db->openfile->has_lock == 2);
+    db->openfile->dirty = 0;
+    return msync(db->openfile->base, len, MS_SYNC|MS_INVALIDATE);
+}
+
+static inline int tm_ensure(struct dbengine *db, size_t offset)
+{
+    struct tm_file *file = db->openfile;
+    if (offset <= file->size) return 0;
+
+    assert(file);
+    assert(file->has_lock == 2);
+
+    offset = tm_roundup(offset);
+
+    // make sure anything we have mapped in has committed to disk
+    if (file->dirty) {
+        if (msync(file->base, file->size, MS_ASYNC)) {
```

Comment:
MS_ASYNC is a no-op on linux, what are we trying to do here (or is this more about portability?)

Reply:
Yeah, that's portability.  Just making sure that anything in the mmap is written down to disk before we unmap it.